### DISCUSSION TITLE
ERROR: An error occurred while attempting to establish an SSL or TLS connection in Postal smtp

### DISCUSSION THREAD

**Discussion by winsta** [postalserver/postal#1458] [Help with using Postal]
can't connect postal SMTP via tls Its shows an error.
Connected to smtp://mydomain:25/?starttls=always
<< 220 mydomain ESMTP Postal/SVYGBU
>> EHLO [172.31.10.74]
<< 250-My capabilities are
<< 250-STARTTLS
<< 250 AUTH CRAM-MD5 PLAIN LOGIN
>> STARTTLS
<< 220 Ready to start TLS
ERROR: An error occurred while attempting to establish an SSL or TLS connection.
This usually means that the SSL certificate presented by the server is not trusted by the system for one or more of
the following reasons:
1. The server is using a self-signed certificate which cannot be verified.
2. The local system is missing a Root or Intermediate certificate needed to verify the server's certificate.
3. A Certificate Authority CRL server for one or more of the certificates in the chain is temporarily unavailable.
4. The certificate presented by the server is expired or invalid.
It is also possible that the set of SSL/TLS protocols supported by the client and server do not match.
See https://github.com/jstedfast/MailKit/blob/master/FAQ.md#SslHandshakeException for possible solutions.

See SMTP server configuration in postal.yml
smtp_server:
  tls_enabled: true
  tls_certificate_path: /etc/letsencrypt/live/domain/fullchain.pem
  tls_private_key_path: /etc/letsencrypt/live/domain/privkey.pem

@willpower232

**Comment by willpower232** [comment]
If you have recently changed the config file, you will need to restart Postal to apply the changes.

Have you changed the ciphers or anything under those lines of config?

Can the postal user definitely access those certificate files you have mentioned?

**Comment by winsta** [comment]
tls working fine but when i test smtp then show me this error

Error: OpenSSL::SSL::SSLError (hostname "myhost.com" does not match the server certificate)

**Comment by rccess** [comment]
run the command 

openssl s_client -connect myhost.com:25 -starttls smtp 

and paste the output

**Comment by winsta** [comment]
CONNECTED(00000005)
depth=0 C = GB, O = Default, OU = Default, CN = default
verify error:num=18:self signed certificate
verify return:1
depth=0 C = GB, O = Default, OU = Default, CN = default
verify return:1
---
Certificate chain
 0 s:C = GB, O = Default, OU = Default, CN = default
   i:C = GB, O = Default, OU = Default, CN = default
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIC/zCCAeegAwIBAgIBADANBgkqhkiG9w0BAQsFADBDMQswCQYDVQQGEwJHQjEQ
MA4GA1UECgwHRGVmYXVsdDEQMA4GA1UECwwHRGVmYXVsdDEQMA4GA1UEAwwHZGVm
YXVsdDAeFw0yMTA3MjEwNDA3MzJaFw0zMTA3MTkwNDA3MzJaMEMxCzAJBgNVBAYT
AkdCMRAwDgYDVQQKDAdEZWZhdWx0MRAwDgYDVQQLDAdEZWZhdWx0MRAwDgYDVQQD
DAdkZWZhdWx0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAz+xXUO1o
A31g0RVmxXWh3G/Q/Ur9Vg6J8xarrOR9iMr5NwlGqfgoUqJVCnCPVCjZykJbRRHQ
KJFv0vvjqHt7qPNWEY6t9wK60YoY+3MV4Fib17ku3xbs1K5obSwCq+INYkQs8YdA
Sndjn5DmFctq/66i8l8QSmh3SBG2QSTFv8+XTV+p40MdPy7OzbCwoizuJIsbx5+2
i+BwdmAplZLOkojHv/HeVihI43onPpKHuUBhNuJWMYs8imQnYd6g9qPXWUWUoujo
5ogmuhiLfKSysaNbHS7LZgpZocOK7DHAV4Ng2n4swiXAt1CCBAwmYhS4lQyME8e9
Ia40+EjM9pjk5wIDAQABMA0GCSqGSIb3DQEBCwUAA4IBAQBT9AhURKhsNRUyRJcQ
OaU6v+YfCEOSJx8kBK9ojKAfUe4jL+tSTxPxidObnYfg8m7dbajM9PzJniJAvtOb
GQAXH3a9VoRGSAtCl/Bsz28KHIvwRsBPJXQcD765LPHBPl0z9uOpEzrGYOm9IzLz
OQTwGNdWHwiSP/D+0MQG89vvJTtWyS0cRIZt1AxPXQBNWSKasSUxo1iYOqibRybm
ptlHVMW/yMH0KA+vcIjytYfkLze07mzSSXPeKDniheLenqwKXD4LjvzstouQeekm
14SWCpn24FJoJa095AWeRLk0ypAfjInok2mHUGJP8SJQEUau22+56Uv9qn2dX1Xs
NWr1
-----END CERTIFICATE-----
subject=C = GB, O = Default, OU = Default, CN = default

issuer=C = GB, O = Default, OU = Default, CN = default

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA
Server Temp Key: DH, 1024 bits
---
SSL handshake has read 1791 bytes and written 553 bytes
Verification error: self signed certificate
---
New, TLSv1.2, Cipher is DHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : DHE-RSA-AES256-GCM-SHA384
    Session-ID: 4395A31B347D85C4661FBF0AA6F786FBD9380104E66E8B061E7FD4276D9D5880
    Session-ID-ctx:
    Master-Key: 375C37BDCD2ADC25442DF56078FEA4A9FEA0963EF59DA627D59C1EA17C330015                                                                                                             34AF34368F645C4B17D5B5391D8FFB57
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - 45 35 cc 7d 82 47 ef 3b-91 58 84 43 5c 5b 7e 87   E5.}.G.;.X.C\[~.
    0010 - 5f f5 15 b6 18 3d 53 5b-a0 62 69 a6 1f 6c f5 eb   _....=S[.bi..l..
    0020 - 74 99 c7 1a 08 71 d6 2a-0c a4 e9 82 ef 0c 64 27   t....q.*......d'
    0030 - 0b f4 ee 06 b0 2f a4 23-f0 e8 b4 58 6e 63 d4 be   ...../.#...Xnc..
    0040 - be cb 72 bf f9 dd 27 72-b1 7a 86 69 73 61 e7 2d   ..r...'r.z.isa.-
    0050 - 92 77 97 49 92 85 f0 97-3c 65 d0 28 c1 dd b6 6e   .w.I....<e.(...n
    0060 - 63 4b d3 97 76 ca f5 7e-72 3e 55 a2 65 86 d6 d7   cK..v..~r>U.e...
    0070 - 72 04 d4 d2 7f 4d dc 89-70 e9 f2 60 a4 2f 49 75   r....M..p..`./Iu
    0080 - 98 ee d3 52 aa 79 4b c0-7d 99 47 ec 34 8e a2 19   ...R.yK.}.G.4...
    0090 - ac fb 53 4e 3b b9 6d c9-09 f4 ed 13 1f e9 fd 36   ..SN;.m........6
    00a0 - 46 9f 94 fd 98 76 1b b3-d5 9a b4 2b 5c d5 5b 68   F....v.....+\.[h
    00b0 - d2 60 d4 cf 32 80 4c c2-83 b3 dc 36 96 d5 56 f7   .`..2.L....6..V.

    Start Time: 1626864436
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
    Extended master secret: no
---
250 AUTH CRAM-MD5 PLAIN LOGIN

**Comment by rccess** [comment]
try to send on 587 for tls

**Comment by winsta** [comment]
140663099892160:error:0200206F:system library:connect:Connection refused:../crypto/bio/b_sock2.c:110:
140663099892160:error:2008A067:BIO routines:BIO_connect:connect error:../crypto/bio/b_sock2.c:111:
connect:errno=111

i get this.

**Comment by rccess** [comment]
try this 

smtp_server:
port: 587
tls_enabled: true
tls_certificate_path: /etc/letsencrypt/live/domain/fullchain.pem
tls_private_key_path: /etc/letsencrypt/live/domain/privkey.pem

**Comment by winsta** [comment]
postal not working if i put port 587

**Comment by rccess** [comment]
check the logs and paste screenshot here of error

**Comment by winsta** [comment]
![Screenshot (77)](https://user-images.githubusercontent.com/34746503/126486353-8761543f-0e65-4f75-9871-18aab56cca03.png)

**Comment by rccess** [comment]
please do send the screenshot of postal.yml



_Note: Please do hide private information in Screenshot or code_

**Comment by willpower232** [comment]
> Can the postal user definitely access those certificate files you have mentioned?

**Comment by aps124** [comment]
I'm also facing the same issue. this is my yml file 


smtp_server:
  default_bind_address: "0.0.0.0"
  port: 587
  default_health_server_port: 9091
  default_health_server_bind_address: 127.0.0.1
  tls_enabled: true
  tls_certificate_path: /opt/postal/config/smtp.cert
  tls_private_key_path: /opt/postal/config/smtp.key
  tls_ciphers:
  ssl_version: SSLv23
  max_message_size: 14

**Comment by rakurtz** [comment]
Hello i am writing this answer because i also struggled to establish a working TLS setup for SMTP. 

These are the steps i took to make it work. 

i copied the the certificates from caddy to the correct folder in postal and changed permissions accordingly:
```
cp /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.eu/postal.mydomain.eu.crt /opt/postal/config/smtp.crt

cp /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.eu/postal.mydomain.eu.key /opt/postal/config/smtp.key

chmod o+r /opt/postal/config/smtp.*
```

after that i altered my postal.yaml to this:

```
smtp_server:
  default_bind_address: "::"
  tls_enabled: true
  tls_certificate_path: /config/smtp.crt
  tls_private_key_path: /config/smtp.key
  tls_ciphers:
  ssl_version: SSLv23
```

**please note:** the path starts with `/`. in some other posts here on GitHub discussions you would find the paths without a beginning slash.

Stop and restarted postal. 

first i encountered some error messages when running this command `openssl s_client -connect postal.mydomain.eu:25 -starttls smtp`:
```
CONNECTED(00000003)
write:errno=104
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 140 bytes and written 353 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

this is what i found in the logs (docker logs postal-smtp-1) so far:
```
2024-07-17 06:51:35 +0000 DEBUG  <= EHLO mail.example.com trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 DEBUG  => 250-My capabilities are trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 DEBUG  => 250-STARTTLS trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 DEBUG  => 250 AUTH CRAM-MD5 PLAIN LOGIN trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 DEBUG  <= STARTTLS trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 DEBUG  => 220 Ready to start TLS trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  An error occurred while processing data from a client. trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  Errno::ENOENT: No such file or directory @ rb_sysopen - config/smtp.crt trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:19:in `read' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:19:in `tls_certificates' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:63:in `ssl_context' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:226:in `block (2 levels) in run_event_loop' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:106:in `select' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:106:in `block in run_event_loop' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:104:in `loop' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:104:in `run_event_loop' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:39:in `block in run' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /usr/local/bundle/gems/klogger-logger-1.4.0/lib/klogger/group_set.rb:18:in `call_without_id' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /usr/local/bundle/gems/klogger-logger-1.4.0/lib/klogger/logger.rb:72:in `tagged' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:37:in `run' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:35 +0000 ERROR  script/smtp_server.rb:13:in `<main>' trace_id=T5O3CJP1 component=smtp-server
2024-07-17 06:51:36 +0000 DEBUG  <= RSET trace_id=UQLPYQ62 component=smtp-server
2024-07-17 06:51:36 +0000 DEBUG  => 250 OK trace_id=UQLPYQ62 component=smtp-server
2024-07-17 06:51:40 +0000 DEBUG  <= AUTH LOGIN trace_id=UQLPYQ62 component=smtp-server
2024-07-17 06:51:40 +0000 DEBUG  => 334 VXNlcm5hbWU6 trace_id=UQLPYQ62 component=smtp-server
```

but after i corrected the the pathes in the `/opt/postal/config/postal.yaml` it worked correctly.

**Reply by plancked** [reply]
hi, what paths did you enter in postal.yaml?

I'm trying:

```
  tls_certificate_path: /config/smtp.crt
  tls_private_key_path: /config/smtp.key
```

or

```
  tls_certificate_path: config/smtp.crt
  tls_private_key_path: config/smtp.key
```

But it does not work:

```
CONNECTED(00000003)
write:errno=104
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 141 bytes and written 354 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

**Reply by rakurtz** [reply]
Hey plancked, i did as i wrote in my little tutorial (with a beginning "/".  
Sorry not to be of more help

**Reply by plancked** [reply]
so certificate files are in /opt/postal/config ?

**Reply by ivansalloum** [reply]
@rakurtz Do you have any idea how much time I spent trying to solve this problem? And to make matters worse, I still don’t understand exactly what went wrong. I followed your exact steps, but the only difference was that I didn’t include these two lines:

```
tls_ciphers:
ssl_version: SSLv23
```

And it didn’t work! I spent the whole day trying various configurations, but I kept getting this error message:

```
2025-03-26 17:43:57 +0000 DEBUG  => 250-STARTTLS trace_id=JBGYWQGC component=smtp-server
2025-03-26 18:49:49 +0000 DEBUG  => 250-STARTTLS trace_id=SSUHW12L component=smtp-server
2025-03-26 18:49:49 +0000 DEBUG  <= STARTTLS trace_id=SSUHW12L component=smtp-server
2025-03-26 18:49:49 +0000 DEBUG  => 220 Ready to start TLS trace_id=SSUHW12L component=smtp-server
2025-03-26 18:49:49 +0000 ERROR  /opt/postal/app/app/lib/smtp_server/server.rb:19:in `tls_certificates' trace_id=SSUHW12L component=smtp-server
```

I tried both a Certbot-generated SSL certificate and a self-signed one, but still no luck. Then, I found your solution, added the two lines you mentioned, and granted read permission for others on both files — and suddenly it works! What a relief! Postal really needs to fix this and automate this configuration for end users. It was a total nightmare. _I did just like you did and now it suddenly works!_

Also, just an extra note: if you’re setting up SSL for mail servers, you should generate a certificate not only for the main domain but also for the MX records. A common cause of the "hostname mismatch" error is when the certificate issued (by Caddy, in my case) for mail.example.com doesn’t include the mail servers (mx1.mail.example.com and mx2.mail.example.com) as Subject Alternative Names (SANs). To fix this, you should issue a certificate that includes all the necessary subdomains.

For example, the certificate should include:

- mail.example.com
- mx1.mail.example.com
- mx2.mail.example.com

You can issue this certificate using Certbot with the following command:

```
sudo certbot certonly --standalone --agree-tos --email hi@gmail.com -d mail.example.com -d mx1.mail.example.com -d mx2.mail.example.com
```

I haven’t tested it fully yet, but I’m testing it now.

**Reply by rakurtz** [reply]
Hey mate, i remember: This was a total nightmare to solve. It seems very unclear and different sources in the internet say different things. I am glad i took the time to write my comment above and was able to help you. Best of luck for you

**Comment by ivansalloum** [comment]
I've written a comprehensive [guide](https://ivansalloum.com/setting-up-postal-as-an-smtp-server/) about setting up Postal. It covers setting up SMTP TLS in detail.

**Reply by paulderkanns** [reply]
@ivansalloum thank you so much for your lovely guide. I also spend quite some time since Sunday because of lack of documentation. Hopefully maintainer will find some time to implement your steps.

May I asked how you connect your mail client? When using Thunderbird it gives me after enable tls:

smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  Connection opened from IP trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  Client identified as IP trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  <= EHLO we-guess.mozilla.org trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 250-My capabilities are trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 250-STARTTLS trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 250 AUTH CRAM-MD5 PLAIN LOGIN trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  <= QUIT trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 221 Closing Connection trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  Connection closed trace_id=YJNVQ2KA component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  Connection opened from 93.208.245.162 trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  Client identified as 93.208.245.162 trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  <= EHLO we-guess.mozilla.org trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 250-My capabilities are trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 250-STARTTLS trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 250 AUTH CRAM-MD5 PLAIN LOGIN trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  <= QUIT trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  => 221 Closing Connection trace_id=NFZSI49B component=smtp-server
smtp-1    | 2025-11-05 10:25:30 +0000 DEBUG  Connection closed trace_id=NFZSI49B component=smtp-server
